src/joystick/n3ds/SDL_sysjoystick.c

```diff
@@ -0,0 +1,367 @@
+
+    SDL_PrivateJoystickAddSensor(joystick, SDL_SENSOR_ACCEL, 0.0f);
+    SDL_PrivateJoystickAddSensor(joystick, SDL_SENSOR_GYRO, 0.0f);
+    SDL_PrivateJoystickAddTouchpad(joystick, 1);
+
+    return 0;
+}
+
+static int
+N3DS_JoystickSetSensorsEnabled(SDL_Joystick *joystick, SDL_bool enabled)
+{
+    sensors_enabled = enabled;
+    return 0;
+}
+
+static void
+N3DS_JoystickUpdate(SDL_Joystick *joystick)
+{
+    N3DSJoystickState previous_state = current_state;
+    hidScanInput();
+
+    UpdateAxis(joystick, &previous_state);
+    UpdateButtons(joystick, &previous_state);
+    UpdateTouch(joystick);
+
+    if (sensors_enabled) {
+        UpdateSensors(joystick, &previous_state);
+    }
+}
+
+SDL_FORCE_INLINE void
+UpdateAxis(SDL_Joystick *joystick, N3DSJoystickState *previous_state)
+{
+    hidCircleRead(&current_state.circlePos);
+    if (previous_state->circlePos.dx != current_state.circlePos.dx) {
+        SDL_PrivateJoystickAxis(joystick,
+                                0,
+                                current_state.circlePos.dx * CORRECTION_FACTOR_X);
+    }
+    if (previous_state->circlePos.dy != current_state.circlePos.dy) {
+        SDL_PrivateJoystickAxis(joystick,
+                                1,
+                                current_state.circlePos.dy * CORRECTION_FACTOR_Y);
+    }
+
+    hidCstickRead(&current_state.cStickPos);
+    if (previous_state->cStickPos.dx != current_state.cStickPos.dx) {
+        SDL_PrivateJoystickAxis(joystick,
+                                2,
+                                current_state.cStickPos.dx * CORRECTION_FACTOR_X);
+    }
+    if (previous_state->cStickPos.dy != current_state.cStickPos.dy) {
+        SDL_PrivateJoystickAxis(joystick,
+                                3,
+                                current_state.cStickPos.dy * CORRECTION_FACTOR_Y);
+    }
+}
+
+SDL_FORCE_INLINE void
+UpdateButtons(SDL_Joystick *joystick, N3DSJoystickState *previous_state)
+{
+    u32 updated_down, updated_up;
+
+    current_state.kDown = hidKeysDown();
+    updated_down = previous_state->kDown ^ current_state.kDown;
+    if (updated_down) {
+        for (Uint8 i = 0; i < joystick->nbuttons; i++) {
+            if (current_state.kDown & BIT(i) & updated_down) {
+                SDL_PrivateJoystickButton(joystick, i, SDL_PRESSED);
+            }
+        }
+    }
+
+    current_state.kUp = hidKeysUp();
+    updated_up = previous_state->kUp ^ current_state.kUp;
+    if (updated_up) {
+        for (Uint8 i = 0; i < joystick->nbuttons; i++) {
+            if (current_state.kUp & BIT(i) & updated_up) {
+                SDL_PrivateJoystickButton(joystick, i, SDL_RELEASED);
+            }
+        }
+    }
+}
+
+SDL_FORCE_INLINE void
+UpdateTouch(SDL_Joystick *joystick)
+{
+    touchPosition touch;
+    Uint8 state;
+    hidTouchRead(&touch);
+    state = (touch.px == 0 && touch.py == 0) ? SDL_RELEASED : SDL_PRESSED;
+
+    SDL_PrivateJoystickTouchpad(joystick,
```

Comment:
I think it makes more sense to expose the touchscreen via the `SDL_Touch` API rather than a joystick touchpad.

Joystick touchpads are indirect touch devices by nature, but your touchscreen has a direct mapping to screen coordinates. Using the touch API will also give you mouse emulation events for free via `SDL_HINT_TOUCH_MOUSE_EVENTS`.